language/diem-vm/src/diem_transaction_executor.rs

```diff
@@ -172,16 +173,43 @@ impl DiemVM {
             cost_strategy
                 .charge_intrinsic_gas(txn_data.transaction_size())
                 .map_err(|e| e.into_vm_status())?;
-            session
-                .execute_script(
+
+            match payload {
+                TransactionPayload::Script(script) => session.execute_script(
                     script.code().to_vec(),
                     script.ty_args().to_vec(),
                     convert_txn_args(script.args()),
                     vec![txn_data.sender()],
                     cost_strategy,
                     log_context,
-                )
-                .map_err(|e| e.into_vm_status())?;
+                ),
+                TransactionPayload::ScriptFunction(script_fn) => session
+                    .execute_script_function(
+                        script_fn.module(),
+                        script_fn.function(),
+                        script_fn.ty_args().to_vec(),
+                        convert_txn_args(script_fn.args()),
+                        vec![txn_data.sender()],
+                        cost_strategy,
+                        log_context,
+                    )
+                    .and_then(|rets| {
+                        if rets.is_empty() {
+                            Ok(())
+                        } else {
+                            // the bytecode verifier should have checked that the script function
+                            // being referred to in the transaction payload has no return values.
+                            Err(
+                                PartialVMError::new(StatusCode::VERIFIER_INVARIANT_VIOLATION)
```

Comment:
@sblackshear @vgao1996, because #7671 allows `execute_script_function` to have return values (but it should not really), we need this extra check here to handle the invariant violation. Just want to make sure that you are OK with the code in this way.

Reply:
Yeah this looks good to me! Although strictly speaking this is more like a VM invariant than a verifier one.

> but it should not really
This is actually a very good point. Right now I think there's still some confusion over the role of script functions and depending on how we clarify that it could go either way. Let's have a separate discussion about it..

Reply:
Yeah this does seem a little bit funky to me. Here's a summary of the current behavior of script functions:
1. Can return values
  a. Can return values to the adapter
2. VM propagates signers for you
3. Signature dynamically checked when called from outside the VM and in this case, return values are not allowed
4. Can be called from another script fun but not a non-script fun

(1) and (3) are both results of (4) but (3) rules out the possibility of (1a) happening at all. So with the current rules, maybe it's better we get rid of the return values from `execute_script_function`? Thoughts @sblackshear?

Although in the future, we should ask ourself
- Should certain rules above be Move requirements or Diem adapter requirements?
- Would it make sense to allow not just script functions, but also scripts to have return values? Would Diem or another adapter be able to use this information?

Reply:
I don't think execute_script_function should be returning values? When was this added? The verifier checks that it doesn't happen (because it is a script)

If you need return values for this adapter it really should be done through execute function. 

You don't want general script entry points returning values because it could violate resource safety. 
I suppose in the future we could open it up if the value has the `Drop` ability, but for now best to keep things simple

Reply:
I pushed for returning values from `execute_script_function` because `script fun`s can return values, but I agree with all of the arguments in favor of removing them. It feels like going forward (as you all suggest):
- A call to a `script fun` that returns values should be an internal call, not made from the outside. A `script fun` that can be called from the outside has the same restrictions as a tx script.
- If a client wants to execute a `script fun` to get return values, it should use `execute_function`

My bad sending you down the wrong path here @vgao1996

Reply:
No worries! I also didn't realize it was impossible for script fun invocations to actually return any values at all. I'll put up a PR to get this removed.

Reply:
Fixed in #7855